### PR TITLE
[Tests-Only]Early Fail Strategy

### PR DESCRIPTION
**Pull request by jasson99**
## Description
This PR enables fail-early system for tests running in CI. 
When ever a test suite fails, it stops the particular build so that the resources are not wasted.
The `drone.star` is adjusted such that for the suites where `earlyFail` feature is set to `True`, any test failure in that suite stops the build. If we do not want the failing tests in a suite to stop the build, we can adjust `earlyFails = False`  for the particular test suite. 

Whenever a test suite fails, a github comment is generated which helps us identify which test failure actually stopped the build.

## Related Issue
- https://github.com/owncloud/QA/issues/672

## Motivation and Context
The drone build seems to be run for all the suites even when a particular suite fails early. To reduce the amount of CI resources used we needed a fail-early system, so that when one pipeline fails the rest of the build is canceled.

## How Has This Been Tested?
- CI

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Technical debt
- [x] Tests

## Checklist:

- [ ] Code changes
- [ ] Unit tests added
- [ ] Acceptance tests added
- [ ] Documentation ticket raised: <link>

### CODE REVIEW

**Comment by ownclouders** [comment]
<details><summary>:boom: Acceptance tests <strong>TrashbinDelete</strong> failed. Please find the screenshots inside ...</summary>

https://drone.owncloud.com/owncloud/web/15713/

<p>


### webUITrashbinDelete-trashbinDelete-feature-113.png
  


</p></details>

**Comment by ownclouders** [comment]
<details><summary>:boom: Acceptance tests <strong>TrashbinDelete</strong> failed. Please find the screenshots inside ...</summary>

https://drone.owncloud.com/owncloud/web/15714/

<p>


### webUITrashbinDelete-trashbinDelete-feature-113.png
  


</p></details>

**Comment by ownclouders** [comment]
<details><summary>:boom: Acceptance tests <strong>TrashbinDelete</strong> failed. Please find the screenshots inside ...</summary>

https://drone.owncloud.com/owncloud/web/15737/

<p>


### webUITrashbinDelete-trashbinDelete-feature-113.png
  


</p></details>

**Comment by ownclouders** [comment]
<details><summary>:boom: Acceptance tests <strong>TrashbinDelete</strong> failed. Please find the screenshots inside ...</summary>

https://drone.owncloud.com/owncloud/web/15744/

<p>


### webUITrashbinDelete-trashbinDelete-feature-113.png
  


</p></details>

**Comment by ownclouders** [comment]
<details><summary>:boom: Acceptance tests <strong>TrashbinDelete</strong> failed. Please find the screenshots inside ...</summary>

https://drone.owncloud.com/owncloud/web/15745/

<p>


### webUITrashbinDelete-trashbinDelete-feature-113.png
  


</p></details>

**Comment by dpakach** [comment]
Looks good, it's working as expected as seen from https://drone.owncloud.com/owncloud/web/15746/24/8
but as of now its difficult to find which pipeline is actually stopping the build. 
I suggest let all the steps run before stopping (eg. upload screenshots etc.) and also make a comment to github so that can find the pipeline that actually failed

File: .drone.star [hunk 1]
```diff
@@ -2053,6 +2054,32 @@ def listRemoteCache():
...
     }]
 
+def stopBuild():
+    return [{
+        "name": "stop-build",
+        "image": "drone/cli:alpine",
+        "pull": "always",
+        "settings": {
+            "server": "https://drone.owncloud.com/",
+            "token": "1efGWn17Riebj5qa4UW4oNaQijtAnONQ",
```
Comment (dpakach) [inline .drone.star:2064]: Please do not expose these tokens in code, you can rather use drone secrets for this.

File: .drone.star [hunk 2]
```diff
@@ -2053,6 +2054,32 @@ def listRemoteCache():
...
+        },
+        "environment": {
+            "DRONE_SERVER": "https://drone.owncloud.com",
+            "DRONE_TOKEN": "1efGWn17Riebj5qa4UW4oNaQijtAnONQ",
```
Comment (dpakach) [inline .drone.star:2068]: same here

File: .drone.star (hunk 1 above)
Comment (jasson99) [reply .drone.star:2064]: Okay. I actually wanted to see if that works. I will make the changes

**Comment by ownclouders** [comment]
<details><summary>:boom: Acceptance tests <strong>TrashbinDelete</strong> failed. Please find the screenshots inside ...</summary>

https://drone.owncloud.com/owncloud/web/15748/

<p>


### webUITrashbinDelete-trashbinDelete-feature-113.png
  


</p></details>

**Comment by ownclouders** [comment]
<details><summary>:boom: Acceptance tests <strong>OCISTrashbinDelete</strong> failed. Please find the screenshots inside ...</summary>

https://drone.owncloud.com/owncloud/web/15753/

<p>


### webUITrashbinDelete-trashbinDelete-feature-106.png
  

### webUITrashbinDelete-trashbinDelete-feature-113.png
  

### webUITrashbinDelete-trashbinDelete-feature-50.png
  

### webUITrashbinDelete-trashbinDelete-feature-86.png
  


</p></details>

**Comment by ownclouders** [comment]
<details><summary>:boom: Acceptance tests <strong>OCISTrashbinDelete</strong> failed. Please find the screenshots inside ...</summary>

https://drone.owncloud.com/owncloud/web/15754/

<p>


### webUITrashbinDelete-trashbinDelete-feature-106.png
  

### webUITrashbinDelete-trashbinDelete-feature-113.png
  

### webUITrashbinDelete-trashbinDelete-feature-50.png
  

### webUITrashbinDelete-trashbinDelete-feature-86.png
  


</p></details>

**Comment by ownclouders** [comment]
<details><summary>:boom: Acceptance tests <strong>OCISTrashbinDelete</strong> failed. Please find the screenshots inside ...</summary>

https://drone.owncloud.com/owncloud/web/15755/

<p>


### webUITrashbinDelete-trashbinDelete-feature-106.png
  

### webUITrashbinDelete-trashbinDelete-feature-113.png
  

### webUITrashbinDelete-trashbinDelete-feature-50.png
  

### webUITrashbinDelete-trashbinDelete-feature-86.png
  


</p></details>

**Comment by ownclouders** [comment]
<details><summary>:boom: Acceptance tests <strong>TrashbinDelete</strong> failed. Please find the screenshots inside ...</summary>

https://drone.owncloud.com/owncloud/web/15755/

<p>


### webUITrashbinDelete-trashbinDelete-feature-113.png
  


</p></details>

**Comment by ownclouders** [comment]
<details><summary>:boom: Acceptance tests <strong>OCISTrashbinDelete</strong> failed. Please find the screenshots inside ...</summary>

https://drone.owncloud.com/owncloud/web/15756/

<p>


### webUITrashbinDelete-trashbinDelete-feature-106.png
  

### webUITrashbinDelete-trashbinDelete-feature-113.png
  

### webUITrashbinDelete-trashbinDelete-feature-50.png
  

### webUITrashbinDelete-trashbinDelete-feature-86.png
  


</p></details>

**Comment by ownclouders** [comment]
<details><summary>:boom: Acceptance tests <strong>OCISTrashbinDelete</strong> failed. Please find the screenshots inside ...</summary>

https://drone.owncloud.com/owncloud/web/15772/

<p>


### webUITrashbinDelete-trashbinDelete-feature-106.png
  

### webUITrashbinDelete-trashbinDelete-feature-113.png
  

### webUITrashbinDelete-trashbinDelete-feature-50.png
  

### webUITrashbinDelete-trashbinDelete-feature-86.png
  


</p></details>

**Comment by ownclouders** [comment]
<details><summary>:boom: Acceptance tests <strong>OCISTrashbinDelete</strong> failed. Please find the screenshots inside ...</summary>

https://drone.owncloud.com/owncloud/web/15773/

<p>


### webUITrashbinDelete-trashbinDelete-feature-106.png
  

### webUITrashbinDelete-trashbinDelete-feature-113.png
  

### webUITrashbinDelete-trashbinDelete-feature-50.png
  

### webUITrashbinDelete-trashbinDelete-feature-86.png
  


</p></details>

File: .drone.star [hunk 3]
```diff
@@ -429,6 +436,8 @@ config = {
             "runningOnOCIS": True,
             "visualTesting": True,
             "filterTags": "not @skip and not @skipOnOCIS and not @notToImplementOnOCIS",
+            "screenShots": True,
```
Comment (Talank) [inline .drone.star:439]: Are we enabling screenshots by default? 🤔

Comment (jasson99) [reply .drone.star:439]: yes

File: .drone.star [hunk 4]
```diff
@@ -1933,6 +1948,9 @@ def runWebuiAcceptanceTests(suite, alternateSuiteName, filterTags, extraEnvironm
             "from_secret": "sauce_access_key",
         }
 
+    if (earlyFail):
+        environment["EARLY_FAIL"] = "true"
```
Comment (individual-it) [inline .drone.star:1952]: why do we need to set the env. variable?

File: .drone.star (hunk 3 above)
Comment (individual-it) [reply .drone.star:439]: but why do we need to change that if it was not set before?

File: .drone.star [hunk 5]
```diff
@@ -800,6 +811,7 @@ def acceptance(ctx):
         "oc10IntegrationAppIncluded": False,
         "skip": False,
         "debugSuites": [],
+        "earlyFail": False,
```
Comment (individual-it) [inline .drone.star:814]: what about setting the default to `True`

**Review by individual-it — CHANGES_REQUESTED** [review]
please explain how that system works in the first post and we also need to document somewhere how to use it

File: .drone.star (hunk 4 above)
Comment (jasson99) [reply .drone.star]: Fixed now

**Comment by jasson99** [comment]
The screenshots for tests running in ocis is set true in this PR so that when ever a build is canceled because of the failure in ocis test suite, the screenshot that is uploaded helps us to debug which test failure actually killed the build.

**Comment by jasson99** [comment]
@individual-it  The `earlyFails` is set to `False` as default, like other keys such as screenshots. If we want the failure of tests of a particular suite to cancel the whole build, we update the `earlyFails` key to `True`. For now, every suite can stop the build, but when we want only certain suites to stop the build later, which might be needed definitely, we can skip the `"earlyFails" = True` in those particular suites.

**Comment by ownclouders** [comment]
<details><summary>:boom: Acceptance tests <strong>webUIOCISRename</strong> failed. Please find the screenshots inside ...</summary>

https://drone.owncloud.com/owncloud/web/15780/

<p>


### webUIRenameFiles-renameFiles-feature-202.png
  

### webUIRenameFiles-renameFiles-feature-227.png
  

### webUIRenameFiles-renameFiles-feature-249.png
  


</p></details>

**Comment by ownclouders** [comment]
<details><summary>:boom: Acceptance tests <strong>OCISTrashbinDelete</strong> failed. Please find the screenshots inside ...</summary>

https://drone.owncloud.com/owncloud/web/15782/

<p>


### webUITrashbinDelete-trashbinDelete-feature-106.png
  

### webUITrashbinDelete-trashbinDelete-feature-113.png
  

### webUITrashbinDelete-trashbinDelete-feature-50.png
  

### webUITrashbinDelete-trashbinDelete-feature-86.png
  


</p></details>

File: .drone.star [hunk 6]
```diff
@@ -107,6 +107,7 @@ config = {
             },
             "visualTesting": True,
             "screenShots": True,
+            "earlyFail": True,
```
Comment (dpakach) [inline .drone.star:110]: You dont need to set this for every one. If you set this to True in default section, then you can only override ones where you need to set false.

File: .drone.star [hunk 7]
```diff
@@ -928,6 +940,9 @@ def acceptance(ctx):
                         if (params["visualTesting"] or (isLocalBrowser(browser) and params["screenShots"])):
                             steps += githubComment()
 
+                        if (params["earlyFail"]):
+                            steps += stopBuild()
```
Comment (dpakach) [inline .drone.star:944]: Also add a GitHub comment step so that we know that the build is being cancelled by some test suite.

**Review by individual-it — APPROVED** [review]

**Comment by ownclouders** [comment]
<details><summary>:boom: Acceptance tests <strong>TrashbinDelete</strong> failed. Please find the screenshots inside ...</summary>

https://drone.owncloud.com/owncloud/web/15802/

<p>


### webUITrashbinDelete-trashbinDelete-feature-113.png
  


</p></details>
<details><summary>:boom: Acceptance tests <strong>TrashbinDelete</strong> failed. The build is cancelled...</summary>

**Comment by jasson99** [comment]
> @individual-it The `earlyFails` is set to `False` as default, like other keys such as screenshots. If we want the failure of tests of a particular suite to cancel the whole build, we update the `earlyFails` key to `True`. For now, every suite can stop the build, but when we want only certain suites to stop the build later, which might be needed definitely, we can skip the `"earlyFails" = True` in those particular suites.

The `earlyFails` has been set to `True` by default now.

File: .drone.star (hunk 5 above)
Comment (jasson99) [reply .drone.star]: fixed

**Comment by ownclouders** [comment]
<details><summary>:boom: Acceptance tests <strong>TrashbinDelete</strong> failed. The build is cancelled...</summary>

**Comment by ownclouders** [comment]
<details><summary>:boom: Acceptance tests <strong>IntegrationApp2</strong> failed. The build is cancelled...</summary>

File: .drone.star [hunk 8]
```diff
@@ -2247,6 +2276,30 @@ def buildGithubComment(suite, alternateSuiteName):
...
+        "pull": "always",
+        "commands": [
+            'echo "<details><summary>:boom: Acceptance tests <strong>%s</strong> failed. The build is cancelled...</summary>\\n\\n" >> /var/www/owncloud/web/comments.file' % alternateSuiteName,
+        ],
+        "environment": {
+            "TEST_CONTEXT": suite,
+            "CACHE_ENDPOINT": {
+                "from_secret": "cache_s3_endpoint",
+            },
+        },
```
Comment (phil-davis) [inline .drone.star:2292]: Are these env vars really needed?

File: .drone.star [hunk 9]
```diff
@@ -1911,7 +1916,7 @@ def copyFilesForUpload():
         ],
     }]
 
-def runWebuiAcceptanceTests(suite, alternateSuiteName, filterTags, extraEnvironment, browser, visualTesting, screenShots):
+def runWebuiAcceptanceTests(suite, alternateSuiteName, filterTags, extraEnvironment, browser, visualTesting, screenShots, earlyFail):
```
Comment (phil-davis) [inline .drone.star:1919]: `earlyFail` is not used in this function. Can it be removed?

Comment (jasson99) [reply .drone.star:1919]: yes.

**Review by xoxys — APPROVED** [review]
LGTM from a drone perspective

**Comment by ownclouders** [comment]
<details><summary>:boom: Acceptance tests <strong>OCISTrashbinDelete</strong> failed. The build is cancelled...</summary>

File: .drone.star (hunk 8 above)
Comment (jasson99) [reply .drone.star]: removed

**Comment by phil-davis** [comment]
It looks like the forced fail is https://drone.owncloud.com/owncloud/web/15840/47/14
The end of the log has:
```
Checking expected failures in /var/www/owncloud/web/tests/acceptance/expected-failures-with-oc10-server-oauth2-login.md
Error: Scenario webUITrashbinDelete/trashbinDelete.feature:113 failed but was not expected to fail.
runsh: Total unexpected failed scenarios throughout the test run:
webUITrashbinDelete/trashbinDelete.feature:113
runsh: There were no unexpected success.
runsh: Exit code: 1
sleeping for 30 seconds at end of test run
```
That is what was expected.

In drone we do not see the build-github-comment* and stop-build steps. I guess that stop-build cancels the whole build before the drone-agent has actually logged the stop-build step. That seems fine.

We did get the github comment "Acceptance tests OCISTrashbinDelete failed. The build is cancelled...". That is good.

@jasson99 this looks good. You could remove the OCISTrashbinDelete test scenario changes, then IMO this is good to approve and merge.

**Comment by jasson99** [comment]
> It looks like the forced fail is https://drone.owncloud.com/owncloud/web/15840/47/14
> The end of the log has:
> 
> ```
> Checking expected failures in /var/www/owncloud/web/tests/acceptance/expected-failures-with-oc10-server-oauth2-login.md
> Error: Scenario webUITrashbinDelete/trashbinDelete.feature:113 failed but was not expected to fail.
> runsh: Total unexpected failed scenarios throughout the test run:
> webUITrashbinDelete/trashbinDelete.feature:113
> runsh: There were no unexpected success.
> runsh: Exit code: 1
> sleeping for 30 seconds at end of test run
> ```
> 
> That is what was expected.
> 
> In drone we do not see the build-github-comment* and stop-build steps. I guess that stop-build cancels the whole build before the drone-agent has actually logged the stop-build step. That seems fine.
> 
> We did get the github comment "Acceptance tests OCISTrashbinDelete failed. The build is cancelled...". That is good.
> 
> @jasson99 this looks good. You could remove the OCISTrashbinDelete test scenario changes, then IMO this is good to approve and merge.

That has been changed now.

**Review by phil-davis — APPROVED** [review]

**Review by LukasHirt — APPROVED** [review]
LGTM :+1:

**Comment by phil-davis** [comment]
Merged. Over the next few days we will see if this is "a good thing" or annoying :)

**Pull request merged**